### ISSUE TITLE
Improving the use of DSPy

### ISSUE THREAD

**Issue by saratavakoli77** [ai-cfia/fertiscan-pipeline#62]
### Context
We are working on refining the prompt and the description of outputs for the language model used in our project. Additionally, we need to revisit the signatures to improve how inputs and outputs are managed, and potentially redesign the flow for complex tasks.

#### 1. Language Model
- [ ] **Improving the prompt**  
  Refine the prompt and provide more details about the task.
  
- [ ] **Improving the description of the output**  
  
- [ ] **Checking if OCR is generating the same text for the same label**  
  Checking if OCR is generating the same text for the same label. Cause using dspy, we have a caching mechanism that will give the same answer if the input is the same. 
  [Link to DSPy Documentation](https://dspy-docs.vercel.app/building-blocks/1-language_models/#:~:text=%3DNone%2C-,cache%3DFalse,-)
  
- [ ] **Evaluate the need for a custom adaptor**  
  Currently, DSPy only offers the `chainOfThoughts` and ReAct adaptor. Discuss whether we need to implement a custom adaptor to better suit our needs, or how to use the existing adaptors.

#### 2. Signature
- [ ] **Review input and output field names**  
  The names of the input and output fields directly influence the prompt and control flow of the language model. Consider whether renaming these fields could improve clarity.

- [ ] ~~**Redesign signature to `context, question -> answer`**~~  
  ~~Consider changing the signature to a format like `context, question -> answer` to simplify the flow.~~
  
- [ ] **Improve the I/O description and docstring**  
  For complex tasks, it's recommended to enrich the input/output descriptions and the docstring at the beginning of the prompt. This can provide a clearer high-level definition of the task.

**Comment by k-allagbe** [comment]
So we're at a "forking" point where on one side, we'll be working on immediate refactoring of the main model and on the other side, on improving the use of DSPy. We probably should not commit both progress on the main branch concurrently.

I suggest the refactoring tasks be allowed to be committed on main while improving the use of DSPy (as a longer research driven endeavour) be committed on a parallel branch created for it. 

Parallel branch is however not easy to manage because:
- we always have to be careful with the destination branch for the PRs
- closing PRs for which the destination branch is not the main doesn't automatically close corresponding issues (might lead to confusion in the project board)

@snakedye @Endlessflow @saratavakoli77 what do you think?

@SonOfLope @ThomasCardin @ChromaticPanic what do you suggest?

**Comment by Endlessflow** [comment]
Pardon my confusion but I am not quite  sure what we mean by refactoring the main model in this context. 

However, I will quickly note that we were already planning to sandbox our experiments to jupiter notebooks. 

DSPy has a lot of tools that are clearly designed with incremental experiments on jupiter notebooks in mind. By going this route we also have the side benefit of making documenting findings easier due to how we can embed markdown blocs in notebooks.

However I suspect this is probably not a fix to the issue you are bringing up @k-allagbe. It's more akin to delaying the issue. It is only a matter of timr before we start getting conclusive experiments and we look to implement the results of those experiments in the pipeline. 

So I suppose my question is what is the time frame for this refactor we are talking about. If it is reasonably short, delaying the implementation of our experiments in the pipeline might be acceptable albeit not ideal.

**Comment by saratavakoli77** [comment]
I agree with James. Let's talk today @k-allagbe

**Comment by k-allagbe** [comment]
@Endlessflow by refactoring the main model I mean changing the FertilizerInspection model to suit our immediate needs.

**Comment by ChromaticPanic** [comment]
Yeah I think we'll need multiple branches and releases for this. Because we may want to preserve previous iterations in a way that lets us compare the results easily.
